packages/dds/merge-tree/Obliterate.md

```diff
@@ -0,0 +1,276 @@
+
+Much of the same logic that goes into conflicting local + remote removal will need to be applied for obliterate.
+Nothing stands out as a conceptual issue or hurdle in this realm, though. Just tricky conditionals.
+
+Once the op is acked, the behavior in the [Remote perspective](#remote-perspective) section suffices for any further concurrent segments.
+
+### Other aspects
+
+TODO: Fill out this section in detail and with as much care as was taken above. For now (draft PR for early feedback), here is my current thinking on the areas I called out:
+
+- Any changes to direct fields of tree nodes themselves (either new data or changes to bookkeeping of existing data)
+> I think `{ obliteratedSeq?: number, localObliteratedSeq?: number }` suffices here. We'll of course need to make sure it's updated where appropriate (segment splits, checks for segment merges, etc.)
+- How the feature interacts with an increasing collab window and zamboni
+> I don't expect many conceptual issues here given the parallels with removedSeq, but need to do a code audit.
+- Impact on the partial lengths scheme
+> This does deserve its own section. Current partial lengths bookkeeping works for removal by adding various adjustment entries based on segments' `removedSeq` and overlapping removes fields. We need to update the bookkeeping on initial obliterate handling as well as on immediate-obliterate-on-inserting-walk cases.
+- Bookkeeping and handling of overlapping removals (note some may be obliterates and some may not be)
+> This was partially covered above. Not sure what else I want to add, but leaving it for now.
+- Reconnection
+> We might need to do fixup on the local state of the merge tree on reconnection by clearing rebased obliterate ops. I don't think the problem is intractible, and can always be done unperformantly in V1 since reconnection of an obliterate should be relatively uncommon.
+- Snapshotting impact
+> We'll need to ensure that obliterated info ends up on segments within the collab window. Otherwise I don't see issues here.
+
+## Endpoint Behavior
+
+One important consideration is what happens near the endpoints of the removed range.
+There are two general possibilities: either the obliterate expands to include segments inserted
+adjacent to the endpoint, or it does't.
+
+```typescript
+
+/** @sealed */
+interface IEndpointMergeStrategy {
+  // TODO
+}
+
+/** @sealed */
+interface IRangeMergeStrategy {
+  start: IEndpointMergeStrategy;
+  end: IEndpointMergeStrategy;
+}
+
+```
+
+## Public API
+
+The public API of sequence will need to be updated for users to leverage the obliterate operation.
+
+```typescript
+class SharedSegmentSequence<TInterval extends IInterval> {
+  public obliterateRange(start: number, end: number, merger: IRangeMergeStrategy = expand)
+}
+```
+
+## Relation to "Move"
+
+TODO: Move is incredibly tricky because unlike removal, it can nest to arbitrarily complex levels. Certain sequences of moves don't even have clear outcomes,
```

Comment:
I've been thinking through `move` a bit more. First, I had incorrectly thought that the semantics SharedTree got were closer to the "sequencing of concurrent ops with the move matters," but in actuality they're closer to obliterate *because all position specifications are relative to existing content*. So if concurrently moving and inserting something nearby a move, if the insert sequences first then the move, the move will resolve the range to include the inserted content. If the move happens first then the insert, the insert will just resolve to the moved place rather than the original place (since it referred to a node inside the removed range).

So, it's not super hard to make well-defined semantics for how a "move" operation ought to work. The existing merge-tree semantics for places (e.g. insertion points) are the same as those of a "place after" reference in SharedTree lingo. For example, if a string starts "ABCDE" and an app does `insert(3, "X")`, it's semantically resolved as though it said "insert X after the "C"".

I need to work through the local bookkeeping, but I think my high-level description of how to implement obliterate can also be extended to the move operation as follows:

- `{ obliteratedSeq, localObliteratedSeq }` instead becomes `{ movedSeq, localMovedSeq }`
- the "move" operation implementation inserts new segments into dst point for each segment in the src range (this may be optimizable to a single segment, need to think about it) and sets appropriate `{ movedSeq, localMovedSeq }` on each segment in the src range. It also sets a field on each src segment which contains a local reference to the corresponding dst segment.
- that local reference field will need some special serialization logic to properly rehydrate it for snapshot purposes
- moved segments act like removed segments for length-counting purposes
- resolving which segment contains a given `{ pos, refSeq, clientId }` in the merge tree will require following the local reference pointers on moved segments

One thing to think about with the "move" operation is how to handle  local references on the existing segment. We also need a good story for interval collections (invalidating the end >= start invariant gets really easy once move is a valid op)

Reply:
i don't think we need to go too deep here, but we should be confident there is a path. in terms of naming, it only matters for persisted things, like ops and snapshots, as in memory naming can be changes quite easily

Reply:
Yeah, I think I'm at that point now. I didn't work through some details of move partial lengths bookkeeping and the excursions in each direction might get more complicated, but it seems doable with enough thought so I'm tabling the issue.